### ISSUE TITLE
Rethink special handling of MT_COLOR_STATUS

### ISSUE THREAD

**Issue by gahr** [neomutt/neomutt#3557]
See #3548 and #3556. This is to rethink/refactor the MT_COLOR_STATUS logic so it doesn't need to be treated differently.

**Comment by chrisdebian** [comment]
Dug into this properly before proposing anything. The special-casing traces to two real structural differences, not just leftover cruft:

1. **Syntax arity** — the generic pattern path (`color/commands.c`) is fixed-arity: `color OBJECT ATTRS FG BG [PATTERN]`, and explicitly errors on any further argument. `color status` needs `[PATTERN [SUBMATCH]]` — an optional *second* argument for the submatch index — which the generic parser can't express without reopening the ambiguity #3548/#3556 already resolved.
2. **Rendering semantics** — every other pattern-capable object colours the whole line on first regex match. Status builds a per-character-range overlay of potentially several overlapping regex matches merged with the base status colour (`dlg_index.c`'s `mutt_draw_statusline`), with its own nearest-start/longest-match tie-breaking. There's no equivalent consumer for the generic list.

So a full merge would mean either (a) extending the generic mechanism to support submatch-based overlay rendering for *all* pattern objects — a real behaviour change beyond just status, or (b) accepting the arity/rendering split is inherent and instead just tidying the parsing front-end so it's one shared entry point with a status-specific branch, rather than status being bounced out at multiple separate call sites the way it is now (`regex.c`, `commands.c`, `dump.c` each have their own special case).

Which direction did you have in mind when this was opened — full unification, or narrowing today's scattered special-casing down to one clean branch point?

**Comment by flatcap** [comment]
I had a play with the Colours a couple of years ago.
(I can push the branches, if you're interested)

---

The Colours are another "global" set that I'd like to break up,
but I'd like a lot of unification in the way they work.
 
Currently, the Colours are a large `enum ColorId` with matching names `struct Mapping ColorFields`.
The two ideas I'd had were:

- Split up the colours
  Create a global `enum ColorDomain` and allow libraries to register their colours.
  Their `ColorId`s would be library-specific.
 
- Add flags
  The parser needs to know which IDs are Simple Colours and which are Regex Colour Arrays.
  (and for Status, which allow Regex backrefs)

---

I had a third, more fiddly, optimisation/unification that I was working on.
The syntax for the Regex Colours allows an "all" pattern, a sort of default, e.g.

```conf
color index_author white default                    # Default colour
color index_author yellow default '~s neomutt'      # Colour Subject of 'neomutt'
```

Currently, the **Default** Colour is stored as a **Simple** Colour,
and the **Pattern** Colours are stored in a separate `RegexColorList`.

I experimented with using a `RegexColorArray`.
Entry **0** was reserved for the **Default** Colour, then the others would be in config order.

**Comment by flatcap** [comment]
As for your "give everything a pattern sub-match" idea, I hadn't considered it before.

At the risk of falling down a rabbit-hole...
Adding it to the Sidebar would add some value - currently the colours are line-based.

**Comment by chrisdebian** [comment]
Yes, I'm interested — please do push those branches, that'd save a lot of re-deriving.

Both ideas (global `ColorDomain` with library-registered `ColorId`s, and flags marking Simple vs Regex-array colours) make sense as a foundation. Happy to start wherever you'd suggest once the branches are up — the `RegexColorArray`-with-reserved-entry-0 experiment sounds like the natural place to pick up if it's already partway there.

On the Sidebar pattern sub-match idea: agreed it's worth adding given colours are currently line-based there, but I'll treat it as a follow-on once the core Colours split is settled rather than scope-creeping the first pass.

**Comment by flatcap** [comment]
Here are my three Colour branches.
Written nearly 2 years ago (pre-AI) and rebased yesterday (hastily).
They compile, but beyond that, no guarantees :-)

- [`colour/domains`](https://github.com/neomutt/neomutt/tree/colour/domains) - Split the colours into library-specific domains
- [`colour/types`](https://github.com/neomutt/neomutt/tree/colour/types) - Flag the types/attributes of the colours
- [`colour/alias`](https://github.com/neomutt/neomutt/tree/colour/alias) - An attempt to add colours to the Alias Dialog

If there's anything useful, take it.

Ideas in my head, at the time:

## Flag the Types/Attributes of Colours

Add flags to each colour definition, telling the parser what to expect.
There are two types:
- **Simple**: Foreground, Background, nothing else
- **Regex**: Foreground, Background, optional Pattern

Plus a flag:
- Regex BackRef (for the Status)

Simple Colours are stored in a big array in `ColorModuleData.simple_colors[]`.

Regex Colours are more complicated.
The "base" Colour, when they have no Pattern, is stored in `simple_colors`.
All "pattern" Colours are stored in a linked-list, `RegexColorList`.

I thought I had a `RegexColorArray`, but I can't find it now.
I used `Array[0]` for the "base" Colour and the other indices, in config order, for the Patterns.

Another similar idea was to use this `RegexColorArray` for the "quoted" Colours.
Currently, they're stored in 10 separate `simple_colors`.

## Split up the colours into domains

This is another wide-scale change.
Though not _necessary_, it follows the same modularisation pattern of the rest of NeoMutt.

The Domain IDs would have a one clear advantage.
In notifications, an observer could use the Domain ID for checking the relevance.

In the branch code, I've given the Colour Domains a new ID.
In the future, the separate Domain IDs could be replaced by the ModuleId.

## Alias Colours

This was a quick hack to introduce colours to the Alias Dialog
(using the `index_*` colours as a template).

- `alias`
- `alias_address`
- `alias_alias`
- ...

So far so good.
The bit where I became tangled up, was the application of Colours.

On each Dialog, there are lots of Colours overlaid and merged (with 'default' being transparent).

For the Index Dialog, Author field, the sequence is:

| Colour         | Type   | Colour for...       |
|:---------------|:-------|:--------------------|
| `normal`       | Simple | All of NeoMutt      |
| `index`        | Simple | All Index Rows      |
| `index`        | Regex  | Matching Index Rows |
| `index_author` | Simple | The Author Column   |
| `index_author` | Regex  | Matching Authors    |

In the `RegexColorArray`, the default entry (0), is effectively a Pattern of `.*` (always).
This would simplify the application of Colours:

- Start with `normal` Colour
- Apply ColourArray: `index`
- Apply ColourArray: `index_author`

---

That's everything I can think of.
Have a poke around, have a think, ask lots of questions.
Feel free to open a discussion.

Cheers!
